crates/ty_python_semantic/src/types/instance.rs

```diff
@@ -40,24 +40,35 @@ impl<'db> Type<'db> {
         M: IntoIterator<Item = (&'a str, Type<'db>)>,
     {
         Self::ProtocolInstance(ProtocolInstanceType::synthesized(
-            SynthesizedProtocolType::new(db, ProtocolInterface::with_members(db, members)),
+            SynthesizedProtocolType::new(
+                db,
+                ProtocolInterface::with_attribute_members(db, members),
+            ),
         ))
     }
 
     /// Return `true` if `self` conforms to the interface described by `protocol`.
-    ///
-    /// TODO: we may need to split this into two methods in the future, once we start
-    /// differentiating between fully-static and non-fully-static protocols.
     pub(super) fn satisfies_protocol(
         self,
         db: &'db dyn Db,
         protocol: ProtocolInstanceType<'db>,
+        relation: TypeRelation,
     ) -> bool {
-        // TODO: this should consider the types of the protocol members
         protocol.inner.interface(db).members(db).all(|member| {
             matches!(
                 self.member(db, member.name()).place,
-                Place::Type(_, Boundness::Bound)
+                Place::Type(ty, Boundness::Bound) if {
+                    let is_todo_type = |ty| {
+                        matches!(
+                            ty,
+                            // TODO: These types have some issues (method signatures are not recognized as fully-static, etc.),
+                            // so they are not checked at the moment.
+                            Type::FunctionLiteral(_)
+                            | Type::PropertyInstance(_)
+                        )
+                    };
+                    is_todo_type(ty) || is_todo_type(member.ty()) || (relation.is_assignability() && ty.is_dynamic()) || ty.is_equivalent_to(db, member.ty())
```

Comment:
If `relation` is `Assignability`, I simply assumed that `ty` satisfies `protocol` only when it is `Dynamic`, but this may not be complete. A class with partially-typed methods in the following code doesn't satisfy the protocol. I think this goes against the gradual guarantee.

https://github.com/mtshiba/ruff/blob/019a1b05008fbb962c354fc31a97ca9b0f89da71/crates/ty_python_semantic/resources/mdtest/protocols.md?plain=1#L563-L571

To fix this, I think we need a check that `ty` can be equivalent to the protocol member type at a materialization
(I think this can be implemented by adding a option such as `MaterializationOption::{Exist, ForAll}` to `is_equivalent_to`)?

Reply:
I pushed some changes to the PR to fix this up. Because of the invariance we need to check `attribute_type.has_relation_to(db, protocol_member_type)` _and_ `protocol_member_type.has_relation_to(db, attribute_type)`